### PR TITLE
Support `DataType::Decimal` for `IN` and `NOT IN` expressions

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

Part of  #2755

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 1]
```diff
@@ -645,8 +744,55 @@ pub fn in_list(
...
+    input_schema: &Schema,
+) -> Result<(Arc<dyn PhysicalExpr>, Vec<Arc<dyn PhysicalExpr>>)> {
+    let expr_type = &expr.data_type(input_schema)?;
+    let list_types: Vec<DataType> = list
+        .iter()
+        .map(|list_expr| list_expr.data_type(input_schema).unwrap())
+        .collect();
+    // TODO in the arrow-rs, should support NULL type to Decimal Data type
+    // TODO support in the arrow-rs, NULL value cast to Decimal Value
+    let result_type = get_coerce_type(expr_type, &list_types);
```
Comment (liukun4515) [inline datafusion/physical-expr/src/expressions/in_list.rs:765]: The IN operator is a shorthand for multiple OR conditions.
We should get the coerced data type for IN expr.

Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:765]: In this issue: https://github.com/apache/arrow-datafusion/issues/2759
The type of list is `BOOLEAN, INT32`, and these two type are not compatible.

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 2]
```diff
@@ -765,35 +909,38 @@ mod tests {
...
+            col_a.clone(),
+            &schema
         );
 
         // expression: "a in (0, 1, NULL)"
         let list = vec![
             lit(ScalarValue::Int64(Some(0))),
             lit(ScalarValue::Int64(Some(1))),
-            lit(ScalarValue::Utf8(None)),
+            lit(ScalarValue::Null),
```
Comment (liukun4515) [inline datafusion/physical-expr/src/expressions/in_list.rs:920]: We can't get the coerced data type between INT64 and UTF8.
I think the type of NULL is  `Null`

Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:920]: From this test case, we can know that we should not treat Null as UTF8 data type

**Comment by liukun4515** [comment]
@alamb @andygrove @viirya  PTAL

File: datafusion/physical-expr/src/planner.rs [hunk 3]
```diff
@@ -310,6 +313,8 @@ pub fn create_physical_expr(
                                 &list_expr_data_type,
                                 &value_expr_data_type,
                             ) {
+                                // TODO: Can't cast from list type to value type directly
```
Comment (liukun4515) [inline datafusion/physical-expr/src/planner.rs:316]: we can't cast the list data type to the value data type directly, and need to coerced common data type if they are compatible.

Comment (liukun4515) [reply datafusion/physical-expr/src/planner.rs:316]: about issue: https://github.com/apache/arrow-datafusion/issues/2759

**Review by yahoNanJing — APPROVED** [review]
LGTM

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 4]
```diff
@@ -300,6 +306,130 @@ fn cast_static_filter_to_set(list: &[Arc<dyn PhysicalExpr>]) -> HashSet<ScalarVa
...
+    negated: bool,
+) -> BooleanArray {
+    let contains_null = list
+        .iter()
+        .any(|v| matches!(v, ColumnarValue::Scalar(s) if s.is_null()));
+    let values = list
+        .iter()
+        .flat_map(|v| match v {
+            ColumnarValue::Scalar(s) => match s {
+                Decimal128(v128op, _, _) => v128op.map(|v128| v128),
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:321]: I don't understand what this `map` does. Is this the same?

```suggestion
                Decimal128(v128op, _, _) => *v128op,
```

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 5]
```diff
@@ -640,13 +785,62 @@ impl PhysicalExpr for InListExpr {
...
+    input_schema: &Schema,
+) -> Result<InListCastResult> {
+    let expr_type = &expr.data_type(input_schema)?;
+    let list_types: Vec<DataType> = list
+        .iter()
+        .map(|list_expr| list_expr.data_type(input_schema).unwrap())
+        .collect();
+    // TODO in the arrow-rs, should support NULL type to Decimal Data type
+    // TODO support in the arrow-rs, NULL value cast to Decimal Value
+    let result_type = get_coerce_type(expr_type, &list_types);
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:813]:
```suggestion
    // https://github.com/apache/arrow-datafusion/issues/2759
    let result_type = get_coerce_type(expr_type, &list_types);
```

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 6]
```diff
@@ -504,6 +634,11 @@ impl PhysicalExpr for InListExpr {
                         .unwrap();
                     set_contains_with_negated!(array, set, self.negated)
                 }
+                DataType::Decimal(_, _) => {
+                    let array = array.as_any().downcast_ref::<DecimalArray>().unwrap();
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:638]: Stylistic suggestion: you can also use https://docs.rs/arrow/16.0.0/arrow/array/fn.as_decimal_array.html here too

```suggestion
                    let array = as_decimal_array(array);
```

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 7]
```diff
@@ -300,6 +306,130 @@ fn cast_static_filter_to_set(list: &[Arc<dyn PhysicalExpr>]) -> HashSet<ScalarVa
...
+                    }
+                }
+            })
+            .collect::<BooleanArray>()
+    } else {
+        // Not in
+        if contains_null {
+            // if the expr is NOT IN and the list contains NULL value
+            // All the result must be NONE
+            vec![None; array.len()]
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:354]: I think using `new_null_array` here might be both simpler and easier to read:

https://docs.rs/arrow/16.0.0/arrow/array/fn.new_null_array.html

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 8]
```diff
@@ -300,6 +306,130 @@ fn cast_static_filter_to_set(list: &[Arc<dyn PhysicalExpr>]) -> HashSet<ScalarVa
...
+                None => None,
+                Some(v128) => {
+                    if values.contains(&v128) {
+                        Some(true)
+                    } else {
+                        Some(false)
+                    }
+                }
+            })
+            .collect::<BooleanArray>()
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:348]:
```suggestion
        array
            .iter()
            .map(|v| v.map(|v128| values.contains(&v128)))
            .collect::<BooleanArray>()
```
Is a bit less verbose and perhaps easier to follow

Comment (alamb) [reply datafusion/physical-expr/src/expressions/in_list.rs:348]: The same could be applied to the `NOT IN` branch

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 9]
```diff
@@ -300,6 +306,130 @@ fn cast_static_filter_to_set(list: &[Arc<dyn PhysicalExpr>]) -> HashSet<ScalarVa
...
+                            Some(false)
+                        }
+                    }
+                })
+                .collect::<BooleanArray>()
+        }
+    }
+}
+
+fn make_set_contains_decimal(
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:375]: The same comments apply here as do `make_list_contains_decimal` as they are largely copy/paste of each other.

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 10]
```diff
@@ -640,13 +785,62 @@ impl PhysicalExpr for InListExpr {
...
+}
+
+fn get_coerce_type(expr_type: &DataType, list_type: &[DataType]) -> Option<DataType> {
+    // get the equal coerced data type
+    list_type
+        .iter()
+        .fold(Some(expr_type.clone()), |left, right_type| {
+            match left {
+                None => None,
+                // TODO refactor a framework to do the data type coercion
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/physical-expr/src/expressions/in_list.rs:840]: Maybe you could start by putting this code in https://github.com/apache/arrow-datafusion/blob/master/datafusion/physical-expr/src/type_coercion.rs or https://github.com/apache/arrow-datafusion/blob/master/datafusion/expr/src/type_coercion.rs 🤔

**Review by alamb — APPROVED** [review]
I left some suggestions for code improvements that I think should be done, but the logic and tests in this PR look good to me. 👍 thanks @liukun4515

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 11]
```diff
@@ -160,6 +165,7 @@ macro_rules! make_contains_primitive {
                 ColumnarValue::Scalar(s) => match s {
                     ScalarValue::$SCALAR_VALUE(Some(v)) => Some(*v),
                     ScalarValue::$SCALAR_VALUE(None) => None,
+                    // TODO this is bug, for primitive the expr list should be cast to the same data type
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:168]: https://github.com/apache/arrow-datafusion/issues/2782

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 4 above)
Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:321]: the type of v128op  is `&Option`, we just need the Some value.
Just like that
```
match v128op {
  None => None
 Some(value) => Some(value)
}
```

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 6 above)
Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:638]: got it.
But i will keep the original style

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 7 above)
Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:354]: I will change the code to 
```
BooleanArray::from(xxx)
```

**Comment by liukun4515** [comment]
All comments have be resolved.

**Review by alamb — APPROVED** [review]
Thank you @liukun4515

**Pull request merged**